src/main/java/com/example/provider/TripMatcher.java

```diff
@@ -109,5 +174,126 @@ public synchronized void triggerMatching() {
         String.format(
             "triggerMatching() completed vehicleId: %s tripName: %s",
             vehicleId, tripToMatch.getName()));
+
+    return true;
+  }
+
+  /**
+   * In case it has been determined that the 'Trip' to match will be part of a 'Shared pool', this
+   * method then is responsible for creating a 'waypoints' list where the given 'Trip' waypoints
+   * will be inserted (position is given by `insertionIndex`).
+   */
+  private static synchronized List<TripWaypoint> getComputedWaypointsForSharedPool(
+      Vehicle vehicle, Trip tripToMatch, String tripToMatchId, int insertIndex) {
+    List<TripWaypoint> vehicleWaypoints = vehicle.getWaypointsList();
+    List<TripWaypoint> computedVehicleWaypoints = new ArrayList<>();
+    computedVehicleWaypoints.addAll(vehicleWaypoints.subList(0, insertIndex));
+
+    computedVehicleWaypoints.add(
+        TripWaypoint.newBuilder()
+            .setLocation(tripToMatch.getPickupPoint())
+            .setTripId(tripToMatchId)
+            .setWaypointType(WaypointType.PICKUP_WAYPOINT_TYPE)
+            .build());
+
+    computedVehicleWaypoints.add(vehicleWaypoints.get(insertIndex));
+
+    List<TerminalLocation> intermediateDestinationsList =
+        tripToMatch.getIntermediateDestinationsList();
+
+    List<TripWaypoint> intermediateDestinationWaypoints =
+        intermediateDestinationsList.stream()
+            .map(
+                location ->
+                    TripWaypoint.newBuilder()
+                        .setLocation(location)
+                        .setTripId(tripToMatchId)
+                        .setWaypointType(WaypointType.INTERMEDIATE_DESTINATION_WAYPOINT_TYPE)
+                        .build())
+            .collect(toList());
+
+    computedVehicleWaypoints.addAll(intermediateDestinationWaypoints);
+
+    computedVehicleWaypoints.add(
+        TripWaypoint.newBuilder()
+            .setLocation(tripToMatch.getDropoffPoint())
+            .setTripId(tripToMatchId)
+            .setWaypointType(WaypointType.DROP_OFF_WAYPOINT_TYPE)
+            .build());
+
+    computedVehicleWaypoints.addAll(
+        vehicleWaypoints.subList(insertIndex + 1, vehicleWaypoints.size()));
+
+    return computedVehicleWaypoints;
+  }
+
+  /**
+   * Calculates the index at where a 'Trip' could be inserted to create a 'Shared pool'.
```

Comment:
Could you add a comment explaining that this is for demonstration purposes and the resulting waypoint list might not be optimal? Actual implementation should take time and distance into account.

Reply:
+1